### ISSUE TITLE
[java] False positive violation of SwitchStmtsShouldHaveDefault when using local variable inference

### ISSUE THREAD

**Issue by vojtapol** [pmd/pmd#2743]
**Affects PMD Version:** 6.26.0

**Description:**

When using local variable inference PMD can't determine that a switch variable is an enum and requires the `default` branch.

**Code Sample demonstrating the issue:**

```java
enum Type {
  CAT, DOG, COW
}

...

for (var animal : getAnimals()) {
    switch (animal.getType()) { // false violation
        case CAT:
            // do something 1
            break;
        case DOG:
            // do something 2
            break;
        case COW:
            // do something 3
            break;
    }
}

...

    public void setValue(List<Type> list) {
        // it works with explicit types: list.forEach((Type eIn) -> {
        list.forEach(eIn -> {
            switch (eIn) {
            case CAT:
                // do something 1
                break;
            case DOG:
                // do something 2
                break;
            case COW:
                // do something 3
                break;
            });
    }


```

**Workaround:**
Either explicit type must be specified in the for-loop or `default` branch must be added to the `switch` statement.

**Comment by oowekyala** [comment]
Thanks for the report! This is a failure in the type resolution code, which will be rehauled in PMD 7. I don't think it's worth fixing on master before then

**Comment by Fernal73** [comment]
When exactly is this mythical version 7 scheduled to ship? 🤔

**Comment by oowekyala** [comment]
There is no scheduled release date as of now. But I hope it's less than a year away.

**Comment by adangel** [comment]
I've added another similar case with lambdas + generics.

**Comment by jsotuyod** [comment]
This is effectively no longer reproducible under PMD 7.0.0
